### ISSUE TITLE
processor clobbers Buffers

### ISSUE THREAD

**Issue by kevinburkeshyp** [balderdashy/sails-postgresql#185]
If you try creating a `'binary'` data type and passing a Buffer as the value, node-postgres returns a Buffer object, but the call to `_.cloneDeep(values)` converts a Buffer into a weird looking object. You can see that in this script: https://gist.github.com/kevinburkeshyp/c45a344a15e040ec6f0e.

I'm not sure how to use that new Buffer object in a meaningful way. Maybe the Buffer should be converted to a string before calling `_.cloneDeep` on it.

**Comment by kevinburkeshyp** [comment]
Here's a possible patch:

```javascript
Processor.prototype.cast = function(table, values) {

  var self = this;
  for (var key in values) {
    if (values.hasOwnProperty(key)) {
      if (Buffer.isBuffer(values[key])) {
        values[key] = values[key].toString('binary');
      }
    }
  }
```

Note now there's a discrepancy between the passed in value (a Buffer) and the result you get back from the database (a string). 

You can't pass in a string, at least, I couldn't get it to work. Creating new values failed with the error message '[object Object]' (that is the literal value, there is no object there) - nothing gets sent to the DB, it fails the anchor validation in waterline. Maybe I should report that as a separate error.

**Comment by particlebanana** [comment]
Shit didn't realize the postgres adapter was doing this. We implemented something like this in waterline here: https://github.com/balderdashy/waterline/commit/e9b618fe0910dd4ffc4a9e4af43a494592c4e9ff looks like it just needs to be moved here.

**Comment by kevinburkeshyp** [comment]
It also does it when trying to pass a Buffer to a find query but I ran out
of time trying to nail it down - was looking in waterline-sequel. Decided
to just use base64 encoded TEXT instead



-kevin

On Aug 19, 2015, at 20:18, Cody Stoltman <notifications@github.com> wrote:

Shit didn't realize the postgres adapter was doing this. We implemented
something like this in waterline here: balderdashy/waterline@e9b618f
<https://github.com/balderdashy/waterline/commit/e9b618fe0910dd4ffc4a9e4af43a494592c4e9ff>
looks like it just needs to be moved here.

—
Reply to this email directly or view it on GitHub
<https://github.com/balderdashy/sails-postgresql/issues/185#issuecomment-132872011>
.

**Comment by sailsbot** [comment]
Thanks for posting, @kevinburkeshyp.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
